### PR TITLE
Support Common output via `BuildDir` and `PublishDir`

### LINKED ISSUE #3497: Support setting a root output path for all projects in a repo or solution with a single property
Currently there are several different properties that control where project output goes.  If you want all project output for a repo to go under a single folder, it is certainly possible to set this up, but it involves setting several properties, and knowing how to include `$(MSBuildProjectName)` as part of the path.  For example, you can put the following in a Directory.Build.props file (assuming all your projects use Microsoft.NET.Sdk):

```xml
<RepoRoot>$([System.IO.Path]::GetFullPath('$(MSBuildThisFileDirectory)'))</RepoRoot>
<BaseOutputPath>$(RepoRoot)artifacts\bin\$(MSBuildProjectName)\</BaseOutputPath>
<BaseIntermediateOutputPath>$(RepoRoot)artifacts\obj\$(MSBuildProjectName)\</BaseIntermediateOutputPath>
```

We would like to add a single property (for example `RootOutputPath`) that handles this automatically.

See also https://github.com/dotnet/sdk/issues/867

### PR DESCRIPTION
**Pull request by Nirmal4G**
Fixes #3497

### Context
We as devs want simplified singular root output path across build and publish. 

### Changes Made
Promote `PublishDir` to `BuildDir` status
Use `BuildDir` to initialize `MSBuildProjectExtensionsPath`
Use `BuildDir` for mismatch warning instead of `BaseIntermediateOutputPath`

### Testing
TBD

### Notes
This is just a concept of the new output paths behaviour using a new common directory to hold all the temporary build outputs. This is not the final implementation as this break previous behaviour. I'm still experimenting, whether to introduce a new property to split the behaviour, either within the targets or to a new targets file with an import switch. If we do favour the spilt, we can do it with #1686. The Sdk way easier since it also reduces duplicating the logic between .NET SDK and the Common targets.

**Please hold up your reviews until it's out of draft.**

### CODE REVIEW

**Comment by Forgind** [comment]
Note for reviewers:
The first five commits are cosmetic; the sixth is the relevant change. Skip to that one.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 1]
```diff
@@ -142,6 +146,8 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     <_OutputPathWasMissing Condition="'$(_OriginalPlatform)' != '' and '$(_OriginalConfiguration)' != '' and '$(OutputPath)' == ''">true</_OutputPathWasMissing>
     <!-- Check whether BaseOutputPath was specified -->
     <BaseOutputPathWasSpecified Condition="'$(BaseOutputPath)' != ''">true</BaseOutputPathWasSpecified>
+    <!-- Initialize BuildDir when empty, to ensure build doesn't produce inconsistent paths -->
+    <BuildDir>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BuildDir)', 'build'))))</BuildDir>
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:150]: If you put `build\` instead of `build`, it'll create one fewer string in the default case.

Also, is that the right default anyway? If we don't want people to have to change anything, and it's currently going to bin, I'd say the default should be empty—that way, it wouldn't change behavior unless the user wants it to.

I glanced at some of the test failures, and the top one was BothBaseOutputPathAndOutputPathWereNotSpecified saying it should be bin, not build\bin.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 2]
```diff
@@ -476,7 +482,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
   <!-- Output location for publish target. -->
   <PropertyGroup>
-    <PublishDir>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(PublishDir)', '$(OutputPath)app.publish'))))</PublishDir>
+    <PublishDir>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(PublishDir)', 'publish'))))</PublishDir>
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:485]: Why this change? Seems like it would confuse people. Did you mean to make it $(BuildDir)app.publish?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 3]
```diff
@@ -839,17 +845,16 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        3. $(BuildDir) is not the same as $(MSBuildProjectExtensionsPath)
 
-      Similar to the error above, there are cases when users set $(BaseIntermediateOutputPath) in the body of their project and things build but only by coincidence.
-      MSBuild does not know if $(BaseIntermediateOutputPath) changing would cause problems so tools like NuGet must set $(EnableBaseIntermediateOutputPathMismatchWarning)
-      to 'true'.
+      Similar to the error above, there are cases when users set $(BuildDir) in the body of their project and things build but only by coincidence.
+      MSBuild does not know if $(BuildDir) changing would cause problems so tools like NuGet must set $(EnableBuildDirMismatchWarning) to 'true'.
     -->
-    <Warning Condition=" '$(EnableBaseIntermediateOutputPathMismatchWarning)' == 'true' And '$(_InitialBaseIntermediateOutputPath)' != '$(BaseIntermediateOutputPath)' And '$(BaseIntermediateOutputPath)' != '$(MSBuildProjectExtensionsPath)' "
+    <Warning Condition=" '$(EnableBuildDirMismatchWarning)' == 'true' And '$(_InitialBuildDir)' != '$(BuildDir)' And '$(BuildDir)' != '$(MSBuildProjectExtensionsPath)' "
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:855]: Where is _InitialBuildDir set?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 4]
```diff
@@ -5474,7 +5479,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
     <RemoveDir
         Directories="$(PublishDir)"
-        Condition="'$(PublishDir)'=='$(OutputPath)app.publish\' and Exists('$(PublishDir)')"/>
+        Condition="'$(PublishDir)' == '$([MSBuild]::EnsureTrailingSlash(`publish`))' and Exists('$(PublishDir)')"/>
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:5482]: I'd prefer just $(BuildDir)app.publish\

**Review by Forgind — CHANGES_REQUESTED** [review]
I'd like to avoid changing any defaults in making this, but I think this is a good start, though I haven't looked to see if I like all the whitespace changes.

@dsplaisted as the issue author.

**Comment by dsplaisted** [comment]
@Nirmal4G Can you remove the whitespace / formatting changes from this PR and submit them separately?  That will make it a lot easier to review the changes related to the output directory.

**Comment by dsplaisted** [comment]
@Nirmal4G Thanks for your work on this.

We try hard to avoid breaking changes in MSBuild.  That means we can't change the output paths used for projects that haven't set `BuildDir`.  It also means that `BaseIntermediateOutputPath` needs to work the same way as it does today, while in this PR it looks like you are replacing its functionality with `BuildDir`.

We will also need to decide whether `BuildDir` is the property name we want to use for this.  I think there were various suggestions, and previously my preference was `RootOutputPath`.  I'm warming up to `BuildDir` with this PR though.  Regardless, I don't think figuring out the name to use should block other progress on this PR.

**Comment by dsplaisted** [comment]
Actually, understanding this more, I don't think it really addresses #3497.  What that issue tracks is a way to set a common output path for multiple projects (ie in a solution/repo).  Then the project name would be automatically appended to that common output path (along with other components, such as `bin` or `obj`, the configuration, platform, TargetFramework, etc.  The way I think we can do that is here: https://github.com/dotnet/msbuild/issues/3497#issuecomment-570268502

I think that's more useful than having a single property that allows you to override both the output and intermediate paths for a single project.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 1 above)
Comment (BenVillalobos) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:150]: I commented this in a different section and I'm seeing that we need to nail down what `BuildDir` is intended to be. https://github.com/dotnet/msbuild/issues/3497#issuecomment-570268502

Daniel suggested that BaseOutputPath and BaseIntermediateOutputPath should be build off of BuildDir. So it sounds to me like BuildDir should just be (by default) the project directory. BOP and BIOP should just default to BuildDir/bin or BuildDir/obj.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 3 above)
Comment (BenVillalobos) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:855]: Line 41 of src/Tasks/Microsoft.Common.props

File: src/Tasks/Microsoft.Common.props [hunk 5]
```diff
@@ -49,10 +49,10 @@ Copyright (C) Microsoft Corporation. All rights reserved.
       management system can write out multiple files but the order of the import is alphabetic because MSBuild sorts the list.
   -->
   <PropertyGroup>
-    <MSBuildProjectExtensionsPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(MSBuildProjectExtensionsPath)', '$(BaseIntermediateOutputPath)'))))</MSBuildProjectExtensionsPath>
+    <MSBuildProjectExtensionsPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(MSBuildProjectExtensionsPath)', '$(BuildDir)'))))</MSBuildProjectExtensionsPath>
```
Comment (BenVillalobos) [inline src/Tasks/Microsoft.Common.props:52]: This was expecting the `obj` folder previously. I'm confused what `BuildDir` is supposed to represent exactly. 

```
    BuildDir:
    Indicates the final output location for the project or solution.
    All the *OutpuPath properties should derive from this.
```

This suggests to me that it would be bin, so this could break whatever takes a dependency on MSBuildProjectExtensionsPath and expects the obj directory.

**Review by BenVillalobos — COMMENTED** [review]
My big concern so far is nailing down exactly what `BuildDir` should be. It sounds like BuildDir should just default to the root of the project? So no matter where it is, BaseOutputPath (bin) and BaseIntermediateOutputPath (obj) just plant themselves at (builddir)/bin or obj respectively.

Also, changing the pre-existing outputpath properties sounds pretty fishy to me. How realistic is it a dev would override these?

Thanks @Forgind for pointing out the 6th commit as the main one to look at.

**Comment by Forgind** [comment]
Before I forget: please retarget to master.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 1 above)
Comment (Nirmal4G) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:150]: @Forgind
> If you put `build\` instead of `build`, it'll create one fewer string in the default case.

True but didn't want to hardcode slashes.

> If we don't want people to have to change anything

This is just one way of doing things, but if we do want to preserve older path spec... we can separate out the older logic and newer logic into a separate props/targets and switch between them with a property say, `UseNewOutputPaths`

@BenVillalobos I thought of that too. My first iteration was the same but this one puts everything under a single folder under project root or solution root.

One of the things we can do with this, is, we can now delete build without publish and vice-versa which many seem to prefer. With this we can also do a force clean which just deletes `build` folder without us having to manually delete `bin`/`obj` folders.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 2 above)
Comment (Nirmal4G) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:485]: No. This was intentional. Making publish a first-class citizen of the MSBuild process. So, I placed it in the root instead.

I've wanted to ask, why the `app.` prefix? why not just `publish`?

Is it because of the explorer's sorting that you wanted to make publish folder come first?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 4 above)
Comment (Nirmal4G) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:5482]: https://github.com/dotnet/msbuild/pull/6105#discussion_r571175158

File: src/Tasks/Microsoft.Common.props [hunk 6]
```diff
@@ -33,12 +33,12 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
   <!--
-      The declaration of $(BaseIntermediateOutputPath) had to be moved up from Microsoft.Common.CurrentVersion.targets
-      in order for the $(MSBuildProjectExtensionsPath) to use it as a default.
+      The $(BuildDir) replaces $(BaseIntermediateOutputPath) to enable common root output path across multiple projects
+      and for the $(MSBuildProjectExtensionsPath) to use it as a default.
   -->
   <PropertyGroup>
-    <BaseIntermediateOutputPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BaseIntermediateOutputPath)', 'obj'))))</BaseIntermediateOutputPath>
-    <_InitialBaseIntermediateOutputPath>$(BaseIntermediateOutputPath)</_InitialBaseIntermediateOutputPath>
+    <BuildDir>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BuildDir)', 'build'))))</BuildDir>
```
[the remark this answers is not in the document]
Comment (Nirmal4G) [reply src/Tasks/Microsoft.Common.props:40]: Yes but if we want to we can keep existing behavior with a property switch.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 7]
```diff
@@ -150,12 +156,12 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
     <ConfigurationName Condition="'$(ConfigurationName)' == ''">$(Configuration)</ConfigurationName>
 
-    <BaseOutputPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BaseOutputPath)', 'bin'))))</BaseOutputPath>
+    <BaseOutputPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BaseOutputPath)', '$(BuildDir)bin'))))</BaseOutputPath>
     <OutputPath Condition="'$(OutputPath)' == '' and '$(PlatformName)' == 'AnyCPU'">$([System.IO.Path]::Combine('$(BaseOutputPath)', '$(Configuration)'))</OutputPath>
     <OutputPath Condition="'$(OutputPath)' == '' and '$(PlatformName)' != 'AnyCPU'">$([System.IO.Path]::Combine('$(BaseOutputPath)', '$(PlatformName)', '$(Configuration)'))</OutputPath>
     <OutputPath>$([MSBuild]::EnsureTrailingSlash('$(OutputPath)'))</OutputPath>
 
-    <BaseIntermediateOutputPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BaseIntermediateOutputPath)', 'obj'))))</BaseIntermediateOutputPath>
+    <BaseIntermediateOutputPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BaseIntermediateOutputPath)', '$(BuildDir)obj'))))</BaseIntermediateOutputPath>
```
[the remark this answers is not in the document]
Comment (Nirmal4G) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:164]: _That is still in progress._ For now, it just takes the value that the user entered but we could take the given path if it's not relative and append to build path if it is.

```suggestion
    <BaseIntermediateOutputPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(BaseIntermediateOutputPath)', 'obj'))))</BaseIntermediateOutputPath>
    <BaseIntermediateOutputPath Condition="'$(BuildDirWasSpecified)' == 'true' AND '$(BuildDir)' != ''">$([System.IO.Path]::Combine('$(BuildDir)', '$(BaseIntermediateOutputPath)'))<BaseIntermediateOutputPath>
```

same for other paths under build/publish folder!

File: src/Tasks/Microsoft.Common.props (hunk 5 above)
Comment (Nirmal4G) [reply src/Tasks/Microsoft.Common.props:52]: Ok, this is a mistake during rebase... silly me!!!

It was supposed to be...

```suggestion
    <MSBuildProjectExtensionsPath>$([MSBuild]::EnsureTrailingSlash($([MSBuild]::ValueOrDefault('$(MSBuildProjectExtensionsPath)', '$(BuildDir)ext'))))</MSBuildProjectExtensionsPath>
```

**Note**
Just like `bin` and `obj`, I've introduced `ext` folder to hold all the project restores, external tools hooking up with msbuild, etc... That way the outputs from these tools are kept separate, since, the pre-build events take a long time in the build process. This way, they are easy to include/copy/ignore in cache-enabled builds (e.g.: docker builds).

**Comment by Nirmal4G** [comment]
@dsplaisted 

> We try hard to avoid breaking changes in MSBuild. That means we can't change the output paths used for projects that haven't set `BuildDir`.

I agree and I do have a solution, same as I did in #5238, we can use a property to switch to a new behaviour when `BuildDir` is set.

> Then the project name would be automatically appended to that common output path

This, I wanted to ask. Do we want to append to `BuildDir` or each of the other paths that depend on (e.g.: `*OutputPath`s)?

1. `build/<project-name>/{bin,obj,ext,...}/...`
2. `build/{bin,obj,ext,...}/<project-name>/...`

Which directory structure do you prefer as the default?

> It also means that `BaseIntermediateOutputPath` needs to work the same way as it does today

I very much prefer if we remove the `MSBuildProjectExtensionsPath`'s dependency on `BaseIntermediateOutputPath`. Anything pre-build goes into the extensions path, so (along with central package versioning feature) caching them is a benefit on builds in the near future. It doesn't hurt to at least plan for it.

Besides, the change I made just relaxes BIOP and not constrain it. So, there'll be no-breaking change on this part but we do have **previews** to check and confirm it.

File: src/Tasks/Microsoft.Common.props (hunk 6 above)
[the remark this answers is not in the document]
Comment (BenVillalobos) [reply src/Tasks/Microsoft.Common.props:40]: I'm not a fan of placing these folders in a subfolder as a default, with an opt out. What's the benefit of this?

**Comment by BenVillalobos** [comment]
To be clear, I'm against changing the default that `bin/` `obj/` folders are located in project roots. If there's an incredibly compelling argument for it I'm all ears, but I imagine this would break a whole lot of people.

If we're talking about simply allowing a user to set `BuildDir`, and `bin/` and `obj/` would be placed at whatever that location is, I see no problems there. 

I'm not sure what the benefit of adding an `ext/` to every build is?

> Just like bin and obj, I've introduced ext folder to hold all the project restores, external tools hooking up with msbuild, etc... That way the outputs from these tools are kept separate, since, the pre-build events take a long time in the build process. This way, they are easy to include/copy/ignore in cache-enabled builds (e.g.: docker builds).

Where would these outputs go previously? in obj?

File: src/Tasks/Microsoft.Common.props (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nirmal4G) [reply src/Tasks/Microsoft.Common.props]: Last paragraph in https://github.com/dotnet/msbuild/pull/6105#discussion_r571171029

**Comment by Nirmal4G** [comment]
> I'm against changing the default...If there's an incredibly compelling argument for it I'm all ears...

#MeToo but yes there's a compelling argument to me is that having a sensible and standard output paths across different types of projects for variety of different use cases. _I'll create a proposal based on what I learn from this PR. We can have the discussion there._

> Where would these outputs go previously? in obj?

Yes, currently but I do propose we move it to a separate folder **only** when under build directory.

**All of this doesn't mean I want to remove existing behavior.**

**Comment by Nirmal4G** [comment]
Closing this since I re-forked the repo, the branch ref to this PR was removed as well. I'll open a new PR soon.
